### PR TITLE
[Tools]: MAE Reconstructed Image Visualization

### PR DESCRIPTION
**Pull request by soonera**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Support MAE reconstructed image visualization

## Modification

(1) Add init_model() function to Initialize a model from checkpoint.
(2) Add inference_model() function to infer an image with the loaded model.
(3) Add forward_test() in Class MAE.
(4) Add unpatchify() in Class MAEPretrainHead.
(5) Add mae_visualization.py and mae_visualization.ipynb

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by soonera** [comment]
The visualization is as follows：

File: tools/misc/mae_visualization.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import torch
+from PIL import Image
+
+from mmselfsup.apis import inference_model, init_model
+
+imagenet_mean = np.array([0.485, 0.456, 0.406])
+imagenet_std = np.array([0.229, 0.224, 0.225])
+
+
+def show_image(image, title=''):
```
Comment (fangyixiao18) [inline tools/misc/mae_visualization.py]: typehint

File: tools/misc/mae_visualization.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    assert image.shape[2] == 3
+    image = torch.clip((image * imagenet_std + imagenet_mean) * 255, 0,
+                       255).int()
+    plt.imshow(image)
+    plt.title(title, fontsize=16)
+    plt.axis('off')
+    return
+
+
+def show_images(x, im_masked, y, im_paste):
```
Comment (fangyixiao18) [inline tools/misc/mae_visualization.py]: typehint

File: mmselfsup/models/algorithms/mae.py [hunk 3]
```diff
@@ -17,7 +22,11 @@ class MAE(BaseModel):
             Defaults to None.
     """
 
-    def __init__(self, backbone=None, neck=None, head=None, init_cfg=None):
+    def __init__(self,
+                 backbone: dict = None,
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/algorithms/mae.py]:
```suggestion
                 backbone: Optional[dict] = None,
```
Please also check other parts

File: mmselfsup/models/algorithms/mae.py [hunk 4]
```diff
@@ -17,7 +22,11 @@ class MAE(BaseModel):
...
+                 neck: dict = None,
+                 head: dict = None,
+                 init_cfg: dict = None) -> None:
         super(MAE, self).__init__(init_cfg)
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/algorithms/mae.py]:
```suggestion
        super().__init__(init_cfg)
```

File: mmselfsup/models/algorithms/mae.py [hunk 5]
```diff
@@ -30,24 +39,23 @@ def __init__(self, backbone=None, neck=None, head=None, init_cfg=None):
     def init_weights(self):
         super(MAE, self).init_weights()
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/algorithms/mae.py]:
```suggestion
        super().init_weights()
```

File: mmselfsup/models/algorithms/mae.py [hunk 6]
```diff
@@ -30,24 +39,23 @@ def __init__(self, backbone=None, neck=None, head=None, init_cfg=None):
...
+    def extract_feat(self, img: torch.Tensor) -> Tuple[torch.Tensor]:
         """Function to extract features from backbone.
 
         Args:
-            img (Tensor): Input images of shape (N, C, H, W).
-
+            img (torch.Tensor): Input images of shape (N, C, H, W).
         Returns:
-            tuple[Tensor]: backbone outputs.
+            tuple[torch.Tensor]: backbone outputs.
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/algorithms/mae.py]:
```suggestion
            Tuple[torch.Tensor]: backbone outputs.
```
If there are items in current data type, better use the capital version. Please check other parts.

File: mmselfsup/models/algorithms/mae.py [hunk 7]
```diff
@@ -30,24 +39,23 @@ def __init__(self, backbone=None, neck=None, head=None, init_cfg=None):
...
         """
         return self.backbone(img)
 
-    def forward_train(self, img, **kwargs):
+    def forward_train(self, img: torch.Tensor,
+                      **kwargs) -> Dict[str, torch.Tensor]:
         """Forward computation during training.
 
         Args:
             img (Tensor): Input images of shape (N, C, H, W).
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/algorithms/mae.py]:
```suggestion
            img (torch.Tensor): Input images of shape (N, C, H, W).
```

File: mmselfsup/models/heads/mae_head.py [hunk 8]
```diff
@@ -18,13 +18,18 @@ class MAEPretrainHead(BaseModule):
         patch_size (int): Patch size. Defaults to 16.
     """
 
-    def __init__(self, norm_pix=False, patch_size=16):
+    def __init__(self, norm_pix: bool = False, patch_size: int = 16) -> None:
         super(MAEPretrainHead, self).__init__()
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/heads/mae_head.py]:
```suggestion
        super().__init__()
```

Comment (YuanLiuuuuuu) [reply mmselfsup/models/heads/mae_head.py]: After fixing this, this PR can be merged.

**Review by YuanLiuuuuuu — APPROVED** [review]

**Pull request merged**